search/includes/classes/class-versioningcleanupjob.php

```diff
@@ -0,0 +1,146 @@
+<?php
+
+namespace Automattic\VIP\Search;
+
+class VersioningCleanupJob {
+
+	/**
+	 * The name of the scheduled cron event to run the versioning cleanup.
+	 */
+	const CRON_EVENT_NAME = 'vip_search_versioning_cleanup';
+
+	const WEEK_IN_SECONDS = 7 * 24 * 60 * 60;
+
+	const SEARCH_ALERT_SLACK_CHAT = '#vip-go-es-alerts';
+	/**
+	 * Initialize the job class.
+	 *
+	 * @access  public
+	 */
+	public function init() {
+		$this->load_dependencies();
+
+		add_action( self::CRON_EVENT_NAME, [ $this, 'versioning_cleanup' ] );
+
+		$this->schedule_job();
+	}
+
+
+
+	protected function load_dependencies() {
+		// Load ElasticPress
+		require_once __DIR__ . '/../../elasticpress/elasticpress.php';
+		$this->indexables = \ElasticPress\Indexables::factory();
+
+		// Index versioning
+		require_once __DIR__ . '/class-versioning.php';
+		$this->versioning = new Versioning();
+	}
+
+	/**
+	 * Schedule class versioning job.
+	 *
+	 * Add the event name to WP cron schedule and then add the action.
+	 */
+	public function schedule_job() {
+		if ( ! wp_next_scheduled( self::CRON_EVENT_NAME ) ) {
+			wp_schedule_event( time(), 'weekly', self::CRON_EVENT_NAME );
+		}
+	}
+
+	/**
+	 * Process versioning cleanup.
+	 */
+	public function versioning_cleanup() {
+		$indexables = $this->indexables->get_all();
+
+		foreach ( $indexables as $indexable ) {
+			$inactive_versions = $this->get_inactive_versions( $indexable );
+
+			foreach ( $inactive_versions as $version ) {
+				if ( ! ( $version['active'] ?? false ) ) {
+					// double check that it is not active
+					$this->send_notification( $indexable->slug, $version['number'] );
+				}
+			}
+		}
+	}
+
+
+	/**
+	 * Retrieve inactive versions to be deleted
+	 *
+	 * @param \ElasticPress\Indexable $indexable The Indexable for which to retrieve index versions
+	 * @return array Array of inactive index versions
+	 */
+	public function get_inactive_versions( \ElasticPress\Indexable $indexable ) {
+		$versions = $this->versioning->get_versions( $indexable );
+
+		if ( ! $versions && ! is_array( $versions ) ) {
+			return [];
+		}
+
+		$active_version = null;
+		foreach ( $versions as $version ) {
+			if ( ( $version['active'] ?? false ) && ( $version['activated_time'] ?? false ) ) {
+				$active_version = $version;
+				break;
+			}
+		}
+
+		if ( ! $active_version ) {
+			return [];
+		}
+
+		$week_ago_breakpoint            = time() - self::WEEK_IN_SECONDS;
+		$was_activated_in_the_last_week = $active_version['activated_time'] > $week_ago_breakpoint;
+		if ( $was_activated_in_the_last_week ) {
+			return [];
+		}
+
+		$inactive_versions = [];
+		foreach ( $versions as $version ) {
+			if ( $version['active'] ?? false ) {
+				continue;
+			}
+
+			if ( ( $version['created_time'] ?? time() ) > $week_ago_breakpoint ) {
```

Comment:
I added this compared to the ticket requirement. This covers cases when we would have active index for a long time but would like to prepare another one to switch to. Without this, we would lose this new version the first time the cron would run.

Reply:
Do we also want to only delete inactive versions that came _before_ the current active version? We could run into situations where version 2 is active, then version 3 is being indexed or migrated too, but that process takes longer than 1 week, at which point the new index is removed even though we still may want to move to it.

If we only consider indexes that have a index number _less than_ the current active index, we can avoid incorrectly deleting indexes that may still be switched to. Thoughts?

Reply:
We decided we won't distinguish between older and newer indexes, we'll just bump the threshold to 1 month instead of 1 week to give a bigger buffer.